### PR TITLE
revise optimizer api and implementation

### PR DESCRIPTION
**Pull request by mattjj**
This new optimizer API is more general, and in addition the new implementation is based on DeviceTuples (which themselves are implemented in #664; the commits are copied here so we'll merge that one first). This is pair work with @skye.

The design goals for the API are are:
1. JAX-compatible in that we can apply any JAX transformations to these functions, with the most important being the application jit and pmap to update_fun
2. Flexibility in optimizer state, particularly removing the current restriction that all parameters have the same flat tuple structure (of equal lengths) and instead allowing optimizer state to have any pytree structure (not just tree product)
3. Allow improved performance by using DeviceTuples, which can avoid pytree traversal overheads on each dispatch

We also updated our implementation of the API to realize the latter advantage.

In the new API, an optimizer is an `(init_fun, update_fun, get_params)` triple of functions, where the component functions have these signatures:

```
init_fun(params)

Args:
  params: pytree representing the initial parameters.

Returns:
  A pytree representing the initial optimizer state, which includes the initial
  parameters and may also include auxiliary values like initial momentum. The
  optimizer state pytree structure generally differs from that of `params`.
```

```
get_params(opt_state)

Args:
  opt_state: pytree representing an optimizer state.

Returns:
  A pytree representing the parameters extracted from `opt_state`, such that
  the invariant `params == get_params(init_params(params))` holds true.
```

```
update_fun(step, grads, opt_state)

Args:
  step: integer representing the step index.
  grads: a pytree with the same structure as `get_params(opt_state)` representing
    the gradients to be used in updating the optimizer state.
  opt_state: a pytree representing the optimizer state to be updated.

Returns:
  A pytree with the same structure as the `opt_state` argument representing the
  updated optimizer state.
```

Notice that an optimizer implementation has a lot of flexibility in the form of `opt_state`: it just has to be a pytree of JaxTypes (so that it can be passed to the JAX transforms defined in api.py) and it has to be consumable by `update_fun` and `get_params`.

The new implementation basically works by flattening pytrees. There are two levels of pytrees to think about: the pytree of `params`, which we can think of as defining an "outer pytree", and a pytree produced by applying `init_fun` to each leaf of the `params` pytree, which we can think of as the "inner pytrees". Since pytrees can be flattened, that structure is isomorphic to a list of lists.

The new optimizer implementation just reifies that list of lists. Moreover it represents the list-of-lists as a JaxTuple-of-JaxTuples so that we can store the entire optimizer state as a single DeviceTuple, and thus pay no tree traversal overhead when we dispatch to a `jit`-compiled `update_fun`. This JaxTuple-of-JaxTuples representing the full optimizer state is stored as the `packed_state` member of the `OptimizerState` namedtuple, which also includes the "outer pytree" definition and the "inner pytree" definitions.

### CODE REVIEW

**Comment by mattjj** [comment]
@skye @hawkinsp many of the commits here are included in #664. To review this one, just look at the changes in jax/experimental/optimizers.py and tests/optimizers_test.py.

**Comment by mattjj** [comment]
Actually @skye one other review-worthy change made here and not in #664 is a revision to `device_get`. On master, `device_get` was something complicated that actually might not work with JaxTuples (since before this change we assumed user code wouldn't use them) or any Tracer instances. Now it looks like this:

```python
device_put = jit(lambda x: x)
device_get = jit(lambda x: x, device_values=False)
```

I added that optional `device_values` argument to `jit`. It just controls whether the output is kept as a device-backed type (xla.DeviceArray or xla.DeviceTuple) or a regular Python type (numpy.ndarray or core.JaxTuple). Before that was only configurable as a global flag (so the logic was already all there, though the only switch was a coarse global), but by adding it as an optional arg to `jit` users can control the behavior more finely, and in particular we can write `device_get` super easily. And now it automatically works with all JaxTypes and even Tracers, because `jit` already does all that work!

Finally, feast your eyes on the deleted code in tree_util.py! And I fixed some super-linear recursive behavior in batching.py but that's probably not worth looking at.

So in summary, these files might be worth looking at:
- jax/experimental/optimizers.py
- corresponding README.md update
- tests/optimizers_test.py
- jax/api.py (just the new `device_get`)
- jax/tree_util.py (just admire the deleted code)

File: jax/api.py [hunk 1]
```diff
@@ -64,7 +64,7 @@
                   "Disable JIT compilation and just call original Python.")
 
 
-def jit(fun, static_argnums=()):
+def jit(fun, static_argnums=(), device_values=True):
```
Comment (skye) [inline jax/api.py]: I think we should be careful about adding keyword arguments like this to the core APIs. It makes the device_get implementation nice, but from an API perspective, it seems odd to have this option when you can call device_get on the outputs. I'm not sure why we have the flag to toggle this behavior either, so maybe I'm missing something, but this doesn't seem useful enough to warrant being part of the jit API.

Comment (mattjj) [reply jax/api.py]: I agree we should be careful, but this is a clear-cut case: it's exactly analogous to how we handle `device_put` on the input side.

Saying we can call `device_get` on outputs is a bit circular, since there is no other `device_get` than this. In other words, this isn't redundant with some other code; quite the opposite, this is the only way to implement `device_get`.

The "`device_get`" on master is implemented differently, but (1) it couldn't handle JaxTuples, ShardedDeviceArrays, or now DeviceTuples or ShardedDeviceTuples (it owuld raise an error), and (2) if it were ever used in a tracing context (e.g. called in a function being differentiated, or in a jitted function) it would raise an error. So in effect `device_get` hasn't been correctly implemented until now.

This new implementation handles all those cases, and with a net reduction in code.

It's pretty clear that this is the right implementation mechanism. Whether we should add the `device_values` argument to the user-facing API, or make it private, is a separate question, and the answer is less clear. But I think it's still better to make it public than keep it private. And it's a Type 2 decision: if we find a clear reason not to support this flag, we can remove its functionality and raise a deprecation warning easily.

Comment (mattjj) [reply jax/api.py]: Would you recommend just adding an underscore, like `_device_values`?

Comment (skye) [reply jax/api.py]: I'd recommend renaming jit to_jit_impl or similar, putting the device_values arg there, and have jit call _jit_impl(device_values=True).

Comment (mattjj) [reply jax/api.py]: Got it. I'll do that so we can decouple the API question entirely, and not have to make any API change here.

Great idea! Thanks for your watchful eye defending our API borders.

**Review by skye — CHANGES_REQUESTED** [review]
I need to review this more, but I think a few high-level comments would help a lot.

File: jax/experimental/optimizers.py [hunk 2]
```diff
@@ -29,35 +29,60 @@
 import jax.numpy as np
 from jax.core import pack
 from jax.util import partial, safe_zip, safe_map, unzip2
-from jax.tree_util import (tree_map, tree_mimomap, tree_structure,
-                           register_pytree_node)
+from jax import tree_util
+from jax.tree_util import tree_flatten, tree_unflatten, register_pytree_node
```
[the remark this answers is not in the document]
Comment (mattjj) [reply jax/experimental/optimizers.py]: Done!

File: jax/experimental/optimizers.py [hunk 3]
```diff
@@ -29,35 +29,60 @@
...
 zip = safe_zip
 
+OptimizerState = collections.namedtuple("OptimizerState",
+                                        ["packed_state", "tree", "subtrees"])
+register_pytree_node(OptimizerState,
+                     lambda xs: ((xs.packed_state,), (xs.tree, xs.subtrees)),
+                     lambda data, xs: OptimizerState(xs[0], data[0], data[1]))
+
 def optimizer(opt_maker):
   """Decorator to make an optimizer map over tuple/list/dict containers."""
```
[the remark this answers is not in the document]
Comment (mattjj) [reply jax/experimental/optimizers.py]: Done!

File: jax/experimental/optimizers.py [hunk 4]
```diff
@@ -134,23 +257,74 @@ def adam(step_size, b1=0.9, b2=0.999, eps=1e-8):
...
+    lst[i:i+1] = x
+    return lst
+
+  def broadcast_into(ndim, x, axis):
+    idx = splice([None] * ndim, axis, [slice(None)])
+    return x[tuple(idx)]
+
+  def init(x0):
+    vs = [np.zeros(sz, dtype=x0.dtype) for sz in x0.shape]
+    return (x0, np.zeros_like(x0), vs)
```
Comment (skye) [inline jax/experimental/optimizers.py]: nit: remove parens

File: jax/experimental/optimizers.py [hunk 5]
```diff
@@ -134,23 +257,74 @@ def adam(step_size, b1=0.9, b2=0.999, eps=1e-8):
...
+
+  def update(i, g, state):
+    x, m, vs = state
+    vs = [broadcast_into(g.ndim, v, i) for i, v in enumerate(vs)]
+    accum = reduce(np.minimum, vs[1:], vs[0]) + g ** 2
```
Comment (skye) [inline jax/experimental/optimizers.py]: I think you can just do reduce(np.minimum, vs) ?

**Review by skye — APPROVED** [review]

File: jax/experimental/optimizers.py [hunk 6]
```diff
@@ -17,7 +17,55 @@
 This short module contains some convenient optimizer definitions, specifically
```
[the remark this answers is not in the document]
Comment (mattjj) [reply jax/experimental/optimizers.py]: Done!

File: jax/experimental/optimizers.py [hunk 7]
```diff
@@ -26,102 +74,177 @@
...
+OptimizerState = collections.namedtuple("OptimizerState",
+                                        ["packed_state", "tree", "subtrees"])
+register_pytree_node(OptimizerState,
+                     lambda xs: ((xs.packed_state,), (xs.tree, xs.subtrees)),
+                     lambda data, xs: OptimizerState(xs[0], data[0], data[1]))
+
 def optimizer(opt_maker):
-  """Decorator to make an optimizer map over tuple/list/dict containers."""
+  """Decorator to make an optimizer defined for arrays generalize to containers.
+
```
[the remark this answers is not in the document]
Comment (mattjj) [reply jax/experimental/optimizers.py]: Done!

File: jax/experimental/optimizers.py [hunk 8]
```diff
@@ -17,7 +17,55 @@
...
+  Returns:
+    A pytree representing the initial optimizer state, which includes the
+    initial parameters and may also include auxiliary values like initial
+    momentum. The optimizer state pytree structure generally differs from that
+    of `params`.
+
+::
+
+  get_params(opt_state)
```
[the remark this answers is not in the document]
Comment (mattjj) [reply jax/experimental/optimizers.py]: Done!

File: jax/experimental/optimizers.py [hunk 9]
```diff
@@ -26,102 +74,177 @@
...
+      return OptimizerState(packed_state, tree, subtrees)
 
-    @functools.wraps(update_fun)
-    def tree_update_fun(i, grad_tree, state_trees):
-      return tree_mimomap(partial(update_fun, i), grad_tree, *state_trees)
+    @functools.wraps(update)
+    def tree_update(i, grad_tree, opt_state):
+      packed_state, tree, subtrees = opt_state
+      grad_flat, tree2 = tree_flatten(grad_tree)
+      assert tree == tree2
```
[the remark this answers is not in the document]
Comment (mattjj) [reply jax/experimental/optimizers.py]: You're right.

File: jax/experimental/optimizers.py [hunk 10]
```diff
@@ -26,102 +74,177 @@
...
+# each leaf of the params pytree, which we can think of as the "inner pytrees".
+# Since pytrees can be flattened, that structure is isomorphic to a list of
+# lists (with no further nesting). This implementation represents that structure
+# as a JaxTuple-of-JaxTuples so that we can maintain the entire optimizer state
+# as a single DeviceTuple, and thus pay no pytree traversal overhead when we
+# dispatch to a `jit`-compiled `update_fun`. That JaxTuple-of-JaxTuples is
+# stored together with the tree structure data in an OptimizerState instance.
+
+OptimizerState = collections.namedtuple("OptimizerState",
+                                        ["packed_state", "tree", "subtrees"])
```
[the remark this answers is not in the document]
Comment (mattjj) [reply jax/experimental/optimizers.py]: Done!

File: jax/experimental/optimizers.py [hunk 11]
```diff
@@ -134,23 +257,74 @@ def adam(step_size, b1=0.9, b2=0.999, eps=1e-8):
...
     x = x - step_size(i) * mhat / (np.sqrt(vhat) + eps)
     return x, m, v
-  return init_fun, update_fun
+  def get_params(state):
+    x, m, v = state
+    return x
+  return init, update, get_params
+
+@optimizer
+def sm3(step_size, momentum=0.9):
```
[the remark this answers is not in the document]
Comment (mattjj) [reply jax/experimental/optimizers.py]: Siiiick!

File: jax/experimental/optimizers.py (hunk 4 above)
Comment (mattjj) [reply jax/experimental/optimizers.py]: Done.

File: jax/experimental/optimizers.py (hunk 5 above)
Comment (mattjj) [reply jax/experimental/optimizers.py]: Seems like you're right!
